src/shims/windows/sync.rs

```diff
@@ -26,17 +30,112 @@ fn srwlock_get_or_create_id<'mir, 'tcx: 'mir>(
+        lock_op: &OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, RwLockId> {
+        let this = self.eval_context_mut();
+        let value_place = this.deref_operand_and_offset(lock_op, 0, this.machine.layouts.u32)?;
+
+        this.rwlock_get_or_create(|ecx, next_id| {
+            Ok(ecx.get_or_create_id(next_id.to_u32_scalar(), &value_place)?.map(RwLockId::from_u32))
+        })
+    }
+
+    // see above: condvars are implemented identically to locks
+    fn condvar_get_or_create_id(
+        &mut self,
+        condvar_op: &OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, CondvarId> {
+        let this = self.eval_context_mut();
+        let value_place = this.deref_operand_and_offset(condvar_op, 0, this.machine.layouts.u32)?;
+
+        this.condvar_get_or_create(|ecx, next_id| {
+            Ok(ecx
+                .get_or_create_id(next_id.to_u32_scalar(), &value_place)?
+                .map(CondvarId::from_u32))
         })
-    })
+    }
+
+    /// Try to reacquire the lock associated with the condition variable after we
+    /// were signaled.
+    fn reacquire_cond_lock(
+        &mut self,
+        thread: ThreadId,
+        lock: RwLockId,
+        shared: bool,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        this.unblock_thread(thread);
+
+        if shared {
+            if this.rwlock_is_locked(lock) {
+                this.rwlock_enqueue_and_block_reader(lock, thread);
+            } else {
+                this.rwlock_reader_lock(lock, thread);
+            }
+        } else {
+            if this.rwlock_is_write_locked(lock) {
+                this.rwlock_enqueue_and_block_writer(lock, thread);
+            } else {
+                this.rwlock_writer_lock(lock, thread);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn wait_on_address_layout(&mut self, size: u64) -> Option<TyAndLayout<'tcx>> {
+        let this = self.eval_context_mut();
+
+        match size {
+            1 => Some(this.machine.layouts.u8),
+            2 => Some(this.machine.layouts.u16),
+            4 => Some(this.machine.layouts.u32),
+            8 => Some(this.machine.layouts.u64),
+            _ => None,
+        }
+    }
+
+    fn read_futex_val(
+        &mut self,
+        ptr_op: &OpTy<'tcx, Tag>,
+        layout: TyAndLayout<'tcx>,
+    ) -> InterpResult<'tcx, u64> {
+        let this = self.eval_context_mut();
+
+        this.check_ptr_access_align(
+            this.read_pointer(ptr_op)?,
+            layout.size,
+            layout.align.abi,
+            CheckInAllocMsg::MemoryAccessTest,
+        )?;
+
+        // FIXME: check if this should be changed when weak memory orders are added.
+        let val = this
+            .read_scalar_at_offset_atomic(ptr_op, 0, layout, AtomicReadOp::SeqCst)?
+            .check_init()?
+            .to_uint(layout.size)?
+            .try_into()
+            .expect("layout should be only u8, u16, u32 or u64 from wait_on_address_layout");
```

Comment:
Why does this support several integer sizes...?

Reply:
WaitOnAddress in windows supports it.